lib/govuk_index/publishing_event_worker.rb

```diff
@@ -56,17 +48,48 @@ def process_action(actions, routing_key, payload)
       else
         logger.info("#{routing_key} -> SKIPPED #{presenter.base_path} #{presenter.type}")
       end
+
+    # Rescuing as we don't want to retry this class of error
+    rescue MissingBasePath => e
+      return if DOCUMENT_TYPES_WITHOUT_BASE_PATH.include?(payload["document_type"])
+      GovukError.notify(e, extra: { message_body: payload })
+        # Unpublishing messages for something that does not exist may have been
+        # processed out of order so we don't want to notify errbit but just allow
+        # the process to continue
+    rescue NotFoundError
+      logger.info("#{payload['base_path']} could not be found.")
+      Services.statsd_client.increment('govuk_index.not-found-error')
+    rescue UnknownDocumentTypeError
+      logger.info("#{payload['document_type']} document type is not known.")
+      Services.statsd_client.increment('govuk_index.unknown-document-type')
     end
 
-    def process_response(response)
-      # we are only expecting to process a single message at a time
-      if response['items'].count > 1
-        Services.statsd_client.increment('govuk_index.elasticsearch.multiple_responses')
-        raise MultipleMessagesInElasticsearchResponse
+    def process_responses(responses, messages)
+      messages_with_error = []
+
+      responses['items'].zip(messages).each do |response, message|
+        messages_with_error << message unless valid_response?(response)
       end
 
-      item = response['items'].first
-      action_type, details = *item.first # item is a hash with a single key, value pair
+      if messages_with_error.any?
+        if messages_with_error.count == messages.count
+          # automatically retry
+          raise(
+            ElasticsearchError,
+            reason: "All requests failed",
+            requests: messages.count,
+            first_request: messages.first,
+            first_error_details: responses['items'].first.first.last
+          )
+        else
+          Services.statsd_client.increment('govuk_index.sidekiq-partial-retry')
+          PublishingEventWorker.perform_async(messages_with_error)
+        end
+      end
+    end
+
+    def valid_response?(response)
+      action_type, details = *response.first # item is a hash with a single key, value pair
```

Comment:
am confused about what this doing. Why is there a `.first` - I thought this was being called with a single item due to `responses['items'].zip(messages).each do |response, message|`

Reply:
response if a hash (if my memory is correct)

```
response = { 'delete' => { details: 'hash'} }
response.first = ['delete', { details: 'hash'}]
```
happy to have a nicer approach to this, but was the best I could think of.

Reply:
Ah, I wasn't aware of `Hash#first` so it looked like you were working with an array.

Does the `*` do anything here though? Seems like you can also write
`action_type, details = response.first`

Reply:
I'm not totally sure I thought it was required, but it may splat automatically. I'll check